### PR TITLE
Activating contravariant subtyping of dependent function types

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** feature

A first prototype in relation with coq/ceps#47.

- [ ] Add support of eta for `lazy`
- [ ] Optimize support of eta for `cbn` and `cbv`
- [ ] Address the lambda-domain conversion issue in `vm`, `native_compute`, `unification`
- [ ] Address the [performance issue](https://github.com/coq/coq/wiki/Performance-experiments#skipping-conversion-of-parameters-of-constructors-and-of-the-domain-of-lambdas-failed) in not converting the domains of lambdas
- [ ] Added / updated test-suite
- [ ] Corresponding documentation was added / updated
- [ ] Entry added in the changelog

### CODE REVIEW

File: kernel/reduction.ml [hunk 1]
```diff
@@ -488,22 +488,22 @@ and eqappr cv_pb l2r infos (lft1,st1) (lft2,st2) cuniv =
            we throw them away *)
         if not (is_empty_stack v1 && is_empty_stack v2) then
           anomaly (Pp.str "conversion was given ill-typed terms (FLambda).");
-        let (x1,ty1,bd1) = destFLambda mk_clos hd1 in
-        let (_,ty2,bd2) = destFLambda mk_clos hd2 in
+        (* We assume terms well-typed in the same supertype, so we don't compare the domains *)
+        let (x1,_ty1,bd1) = destFLambda mk_clos hd1 in
+        let (_,_ty2,bd2) = destFLambda mk_clos hd2 in
```
Comment (SkySkimmer) [inline kernel/reduction.ml:493]: Sadly in the past this killed performance https://github.com/coq/coq/wiki/Performance-experiments#skipping-conversion-of-parameters-of-constructors-and-of-the-domain-of-lambdas-failed

Comment (herbelin) [reply kernel/reduction.ml:493]: Ah, ok. So this means that we may require a light conversion (where all sorts are identified)?

Comment (SkySkimmer) [reply kernel/reduction.ml]: Maybe.
I guess that would look like 547448a584c06c682e317103ac6f5bf4cf05f75d (pushed on top of this PR)

**Comment by ppedrot** [comment]
This would require a paper in a high-profile conference to justify the very beginning. I don't see that anywhere, and I can't imagine this PR being merged in any reasonable timeframe, so no point in keeping it open.

**Comment by herbelin** [comment]
How is it different from the current situation? Bruno's model uses typed conversion and iirc relies on SN to show that untyped conversion can be typed. From the moment we use typed conversion, interpreting eta and contravariant subtyping is not a problem.

In any case, I think this is something to work on (even if not me). The theory with contravariant subtyping and eta has nicer properties.

**Comment by herbelin** [comment]
One more comment: Bruno [defines](https://github.com/barras/cic-model/blob/dcc38f3104048aa50d230f819085131b16702d3d/ZFrelations.v#L947-L949) the set-theoretic interpretation of `forall x:T, U` as the graph of functions with domain the interpretation of `T` and codomain included in the interpretation of `U`. This does not support contravariance but we could probably define it instead as the set of pairs of a superset of the interpretation of `T` and of the graph of a function with domain this superset. This would then satisfy contravariance.

Would an adaptation of the implementation of Bruno to this alternative definition be convincing enough? (I'm not going to do it but it might be an interesting challenge.)

**Comment by herbelin** [comment]
Yet a comment: Presentations of type theory using type-directed conversion and eta-expansion on the fly have no problems to handle contravariant subtyping. The current restriction to "Luo's system" is probably just an historical accident coming from the combination of using eta-free untyped conversion + relying only on the standard (non-contravariant) interpretation of dependent product types in set theory.

**Pull request closed without merging**